aleph-client/src/treasury.rs

```diff
@@ -0,0 +1,164 @@
+        treasury_era_payout_from_staking
+    );
+    treasury_era_payout_from_staking
+}
+
+/// Creates a proposal of spending treasury's funds.
+///
+/// The intention is to transfer `value` balance to `beneficiary`. The signer of `connection` is the
+/// proposer.
+pub fn propose(
+    connection: &SignedConnection,
+    value: Balance,
+    beneficiary: &AccountId32,
+) -> ApiResult<Option<H256>> {
+    let xt = compose_extrinsic!(
+        connection.as_connection(),
+        "Treasury",
+        "propose_spend",
+        Compact(value),
+        GenericAddress::Id(beneficiary.clone())
+    );
+    try_send_xt(connection, xt, Some("treasury spend"), XtStatus::Finalized)
+}
+
+#[derive(Debug, Decode, Copy, Clone)]
+struct ProposalRejectedEvent {
+    proposal_id: u32,
+    _slashed: Balance,
+}
+
+fn wait_for_rejection<C: AnyConnection>(connection: &C, proposal_id: u32) -> AnyResult<()> {
+    wait_for_event(
+        connection,
+        ("Treasury", "Rejected"),
+        |e: ProposalRejectedEvent| {
+            info!("[+] Rejected proposal {:?}", e.proposal_id);
+            proposal_id.eq(&e.proposal_id)
+        },
+    )
+    .map(|_| ())
+}
+
+fn send_rejection(connection: &RootConnection, proposal_id: u32) -> ApiResult<Option<H256>> {
+    let xt = compose_extrinsic!(
+        connection.as_connection(),
+        "Treasury",
+        "reject_proposal",
+        Compact(proposal_id)
+    );
+    try_send_xt(
+        connection,
+        xt,
+        Some("treasury rejection"),
+        XtStatus::Finalized,
+    )
+}
+
+/// Rejects proposal with id `proposal_id` and waits for the corresponding event.
+///
+/// Fails if either sending extrinsic failed or the expected event has not been observed.
+pub fn reject(connection: &RootConnection, proposal_id: u32) -> AnyResult<()> {
+    let listener = {
+        let (c, p) = (connection.clone(), proposal_id);
+        thread::spawn(move || wait_for_rejection(&c, p))
+    };
+    send_rejection(connection, proposal_id)?;
+    listener
+        .join()
+        .expect("Corresponding event should have been emitted")
+}
+
+fn send_approval(connection: &RootConnection, proposal_id: u32) -> ApiResult<Option<H256>> {
+    let xt = compose_extrinsic!(
+        connection.as_connection(),
+        "Treasury",
+        "approve_proposal",
+        Compact(proposal_id)
+    );
+    try_send_xt(
+        connection,
+        xt,
+        Some("treasury approval"),
+        XtStatus::Finalized,
+    )
+}
+
+fn wait_for_approval<C: AnyConnection>(connection: &C, proposal_id: u32) -> AnyResult<()> {
+    loop {
+        let approvals: Vec<u32> = connection
+            .as_connection()
+            .get_storage_value("Treasury", "Approvals", None)
+            .expect("Key `Treasury::Approvals` should be present in storage")
+            .unwrap_or_default();
+        if approvals.contains(&proposal_id) {
+            info!("[+] Proposal {:?} approved successfully", proposal_id);
```

Comment:
Comment: same here, these logs seem out of place to me in a client lib.

Reply:
I removed all logs in this module